runtime/compiler/control/CompileBeforeCheckpoint.cpp

```diff
@@ -0,0 +1,124 @@
+   _region(region),
+   _vmThread(vmThread),
+   _fej9(fej9),
+   _compInfo(compInfo),
+   _methodsSet(std::less<TR_OpaqueMethodBlock*>(), _region)
+   {}
+
+void
+TR::CompileBeforeCheckpoint::addMethodToList(TR_OpaqueMethodBlock *method)
+   {
+   if (method && _methodsSet.find(method) == _methodsSet.end())
+      {
+      _methodsSet.insert(method);
+      }
+   }
+
+void
+TR::CompileBeforeCheckpoint::queueMethodsForCompilationBeforeCheckpoint()
+   {
+   for (auto iter = _methodsSet.begin(); iter != _methodsSet.end(); iter++)
+      {
+      TR_OpaqueMethodBlock *method = *iter;
+      J9Method *j9method = reinterpret_cast<J9Method *>(method);
+
+      J9ROMMethod *romMethod = J9_ROM_METHOD_FROM_RAM_METHOD(j9method);
+      if (
+          /* Method is interpreted */
+          !_compInfo->isCompiled(j9method)
+
+          /* Method is not a thunk archetype */
+          && !_J9ROMMETHOD_J9MODIFIER_IS_SET(romMethod, J9AccMethodFrameIteratorSkip)
+
+          /* Method is not abstract */
+          && !_J9ROMMETHOD_J9MODIFIER_IS_SET(romMethod, J9AccAbstract)
+         )
+         {
+         if (TR::Options::getCmdLineOptions()->getVerboseOption(TR_VerboseCheckpointRestoreDetails))
+            {
+            TR_VerboseLog::CriticalSection cs;
+            TR_VerboseLog::write(TR_Vlog_CHECKPOINT_RESTORE, "Attempting to queue ");
+            _compInfo->printMethodNameToVlog(j9method);
+            TR_VerboseLog::writeLine(" (%p) for compilation", j9method);
+            }
+
+         TR_MethodEvent event;
+         event._eventType = TR_MethodEvent::CompilationBeforeCheckpoint;
+         event._j9method = j9method;
+         event._oldStartPC = 0;
+         event._vmThread = _vmThread;
+         event._classNeedingThunk = 0;
+         bool newPlanCreated;
+         TR_OptimizationPlan *plan = TR::CompilationController::getCompilationStrategy()->processEvent(&event, &newPlanCreated);
+         // the plan needs to be created only when async compilation is possible
+         // Otherwise the compilation will be triggered on next invocation
+         if (plan)
+            {
+            bool queued = false;
+
+               // scope for details
+               {
+               TR::IlGeneratorMethodDetails details(j9method);
+               IDATA result = (IDATA)_compInfo->compileMethod(_vmThread, details, NULL, TR_maybe, NULL, &queued, plan);
+               }
+
+            if (!queued && newPlanCreated)
+               TR_OptimizationPlan::freeOptimizationPlan(plan);
+            }
+         }
+      }
+   }
+
+void
+TR::CompileBeforeCheckpoint::collectAndCompileMethodsBeforeCheckpoint()
+   {
+   /* Acquire Class Unload Monitor to prevent class unloading */
+   TR::MonitorTable::get()->getClassUnloadMonitor()->enter_write();
+
+   collectMethodsForCompilationBeforeCheckpoint();
+   queueMethodsForCompilationBeforeCheckpoint();
+
+   TR::MonitorTable::get()->getClassUnloadMonitor()->exit_write();
```

Comment:
If an exception is thrown in one of those 2 methods above, but not caught, the class unload monitor may remain locked.
Since `collectMethodsForCompilationBeforeCheckpoint();` is empty at this point, it's hard to guess what will happen in the future.Maybe we should catch any exception here, release the monitor, then rethrow.

Reply:
To keep the code clean, I just added a new critical section class to enter/exit the class unload monitor appropriately in https://github.com/eclipse-openj9/openj9/compare/2b177dd0f99d9e9ea49c7b4599d1972b1e89cd93..534463ab114cd1c08395400c07b98fb9bbafc826